### ISSUE TITLE
The density_for_timestep_sampling and loss_weighting for SD3 Training！！！

### ISSUE THREAD

**Issue by DidiD1** [huggingface/diffusers#9056]
Thanks to Rafie Walker's code we can try to train SD3 models with flow-matching! 
But some places don't seem to match what's in the paper.
Rafie Walker's code is below:
```
def compute_density_for_timestep_sampling(
    weighting_scheme: str, batch_size: int, logit_mean: float = None, logit_std: float = None, mode_scale: float = None
):
    if weighting_scheme == "logit_normal":
        # See 3.1 in the SD3 paper ($rf/lognorm(0.00,1.00)$).
        u = torch.normal(mean=logit_mean, std=logit_std, size=(batch_size,), device="cpu")
        u = torch.nn.functional.sigmoid(u)
    elif weighting_scheme == "mode":
        u = torch.rand(size=(batch_size,), device="cpu")
        u = 1 - u - mode_scale * (torch.cos(math.pi * u / 2) ** 2 - 1 + u)
    else:
        u = torch.rand(size=(batch_size,), device="cpu")
    return u

def compute_loss_weighting_for_sd3(weighting_scheme: str, sigmas=None):
    if weighting_scheme == "sigma_sqrt":
        weighting = (sigmas**-2.0).float()
    elif weighting_scheme == "cosmap":
        bot = 1 - 2 * sigmas + 2 * sigmas**2
        weighting = 2 / (math.pi * bot)
    else:
        weighting = torch.ones_like(sigmas)
    return weighting
```
 My question is below:

1. when weighting_scheme == "mode, the code only compute the f_mode. If you need to compute 'u', you should some additional operation？
2. Cos-map seems to compute the weight of timesteps, not the weight of loss?
3. when we use logit_normal, it based on the RF-setting. So the weight of the loss should be t/(1-t), but the code doesn't compute the weight instead of torch.ones_like(sigmas)?

So I think there need some modify to correct compute the loss of SD3!
Thanks for discussion together!

**Comment by bghira** [comment]
honestly none of the weighting tricks really seem relevant to finetuning SD3. not using the timestep weighting has better results.

**Comment by xiao2mo** [comment]
> honestly none of the weighting tricks really seem relevant to finetuning SD3. not using the timestep weighting has better results.

Could u give some more details, thanks a lot

**Comment by bghira** [comment]
yes, if you look at the timestep selection distribution using the SD3 style training, it effectively does not ever train the 900-1000 or 0-100 range of timesteps. they are just ignored:

![image](https://github.com/user-attachments/assets/6f622e20-3bbb-46a3-aeb1-8f9794d2f8ea)

ignoring the gaps in the chart here (wandb was having issues) the timestep selection at the end is where i switched to uniform sampling and the model started learning composition and details properly

**Comment by DidiD1** [comment]
This phenomenon was mentioned in the SD3 paper，maybe why they proposed 'mode sampling with heavy-tails' time-sampling method. However it's strange that in their experiment results 'log-norm' is much better the 'mode' and uniform sampling. So I guess that maybe the different sampling method has their special advantages which needs experiment to valid which one is suitable for own task.

**Comment by DidiD1** [comment]
This phenomenon was mentioned in the SD3 paper，maybe why they proposed 'mode sampling with heavy-tails' time-sampling method. However it's strange that in their experiment results 'log-norm' is much better the 'mode' and uniform sampling. So I guess that maybe the different sampling method has their special advantages which needs experiment to valid which one is suitable for own task.

**Comment by bghira** [comment]
![image](https://github.com/user-attachments/assets/fe335f61-63dc-4514-8c72-98b7bc72de60)

**Comment by bghira** [comment]
it just needs an absolutely enormous batch size for these to make sense.

**Comment by DidiD1** [comment]
Thanks a lot. And for my question3:
"when we use logit_normal, it based on the RF-setting. So the weight of the loss should be t/(1-t), but the code doesn't compute the weight instead of torch.ones_like(sigmas)?"

**Comment by xiao2mo** [comment]
> yes, if you look at the timestep selection distribution using the SD3 style training, it effectively does not ever train the 900-1000 or 0-100 range of timesteps. they are just ignored:
> 
> ![image](https://private-user-images.githubusercontent.com/59658056/355660005-6f622e20-3bbb-46a3-aeb1-8f9794d2f8ea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQyMjUzNjYsIm5iZiI6MTcyNDIyNTA2NiwicGF0aCI6Ii81OTY1ODA1Ni8zNTU2NjAwMDUtNmY2MjJlMjAtM2JiYi00NmEzLWFlYjEtOGY5Nzk0ZDJmOGVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIxVDA3MjQyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA0MmZlY2Q1Yzc2NmM0ZTc0NjI2ZDZiNTA5Zjg2M2MyYzIyY2I2YTRhYzRkZTZkYzFiOTZlOTVmMjFkZDY2ZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.L2u2G9GybvABl_LGjZYMbz2VUI5jaxHaE6nkA2LQ3TE)
> 
> ignoring the gaps in the chart here (wandb was having issues) the timestep selection at the end is where i switched to uniform sampling and the model started learning composition and details properly

Thanks a lot

**Comment by ivylilili** [comment]
> yes, if you look at the timestep selection distribution using the SD3 style training, it effectively does not ever train the 900-1000 or 0-100 range of timesteps. they are just ignored:
> 
> ![image](https://private-user-images.githubusercontent.com/59658056/355660005-6f622e20-3bbb-46a3-aeb1-8f9794d2f8ea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQzOTQ1MzMsIm5iZiI6MTcyNDM5NDIzMywicGF0aCI6Ii81OTY1ODA1Ni8zNTU2NjAwMDUtNmY2MjJlMjAtM2JiYi00NmEzLWFlYjEtOGY5Nzk0ZDJmOGVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIzVDA2MjM1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdkZTE2NjczMDk5YjQ3MDk0MzBiZDFiMDA0NGFmMzI2NDRhZjMyMTJhZjYzY2E5YTVkMzU3NTU0MzM2NzEzMDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tpCW-rFaAKBqChifV8iYcZXRrqN9WX-dnj5ibdVJHpk)
> 
> ignoring the gaps in the chart here (wandb was having issues) the timestep selection at the end is where i switched to uniform sampling and the model started learning composition and details properly

@bghira Hi bghira~ I'd like to know when you try the  "SD3 style training (lognorm sampling)" or "uniform sampling", what is the difference between the training loss?  When you switched to uniform sampling, did it help to lower the loss curve? In my uniform training, these is still some artifacts in the generated image, so I wonder which part in the noise sampling is important to improve this problem?  Want to hear your insights, Thanks~

**Comment by bghira** [comment]
currently we're using sigmoid sampling for timesteps which seems fine but no one has really ablated whether it leaves fine details out

**Comment by culeao** [comment]
> currently we're using sigmoid sampling for timesteps which seems fine but no one has really ablated whether it leaves fine details out

Actually, sigmoid and lognorm are mathematically equivalent. But I'm curious why existing open source training implementations don't use timeshift during training, but SD3 paper does.

**Comment by DidiD1** [comment]
> > currently we're using sigmoid sampling for timesteps which seems fine but no one has really ablated whether it leaves fine details out
> 
> Actually, sigmoid and lognorm are mathematically equivalent. But I'm curious why existing open source training implementations don't use timeshift during training, but SD3 paper does.
> 
> <img alt="image" width="587" src="https://private-user-images.githubusercontent.com/35450998/361637568-d8708fdf-cf82-4a53-89a3-0a729b10e258.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ3NTAwOTAsIm5iZiI6MTcyNDc0OTc5MCwicGF0aCI6Ii8zNTQ1MDk5OC8zNjE2Mzc1NjgtZDg3MDhmZGYtY2Y4Mi00YTUzLTg5YTMtMGE3MjliMTBlMjU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI3VDA5MDk1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ1ZTYwYTMxNTFkM2YwODY1ZWJkZTgyOTY1NTkxMWQ3NzU2Zjc3ZjgxNjRkNGQ2N2Y1NGU1NWYzOTMzMTVlMmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.auq9VDWI9SfSBuBR0gzaYqDJyWYUJazwNRRzGEtLxxc">

In fact, the diffusers version for SD3 has used the timashifting, You can see it in the init of FlowMatchEulerDiscreteScheduler,
{
  "_class_name": "FlowMatchEulerDiscreteScheduler",
  "_diffusers_version": "0.29.0.dev0",
  "num_train_timesteps": 1000,
  "shift": 3.0
}

# shcedule
sigmas = shift * sigmas / (1 + (shift - 1) * sigmas)

**Comment by jjihwan** [comment]
> Thanks a lot. And for my question3: "when we use logit_normal, it based on the RF-setting. So the weight of the loss should be t/(1-t), but the code doesn't compute the weight instead of torch.ones_like(sigmas)?" Do I need to modify the loss weight？

@DidiD1 Hi! Thank you for your valuable discussion about timestep schedule. It really helps me. BTW do you find any answer of this question?

**Comment by wgsxm** [comment]
> Thanks a lot. And for my question3: "when we use logit_normal, it based on the RF-setting. So the weight of the loss should be t/(1-t), but the code doesn't compute the weight instead of torch.ones_like(sigmas)?" Do I need to modify the loss weight？

Thank you for your insightful discussion. Do you find any answer to the question? If training if fine with torch.ones_like(sigmas), does this mean that the loss weighting doesn't make much difference to the result?

**Comment by CyberPegasus** [comment]
Re-sampling timesteps has the similar effects as changing the weight of loss.
